src/Components/src/Microsoft.AspNetCore.Components.Server/Circuits/RemoteRenderer.cs

```diff
@@ -101,14 +101,17 @@ public void Dispose()
         protected override void AddToRenderQueue(int componentId, RenderFragment renderFragment)
         {
             // Render operations are not thread-safe, so they need to be serialized.
-            // This also ensures that when the renderer invokes component lifecycle
-            // methods, it does so on the expected sync context.
-            // We have to use "Post" (for async execution) because if it blocked, it
-            // could deadlock when a child triggers a parent re-render.
-            _syncContext.Post(_ =>
+            if (SynchronizationContext.Current != _syncContext)
```

Comment:
Do we know the cases where this happens? Did **user code** initiate a background operation? Or do we have a bug that leads to `AddToRenderQueue` being called from outside the sync context?

If this is happening because **we are** running operations outside of the sync context then these are bugs that need to be fixed 😆 

If user code did this (let's say you set up a timer) then this could be OK if the **only** thing user code is doing is triggering a render. 

Let's suppose you've got a TODO list, and you're checking a webservice at intervals to update the list. Inside that TODO list you have a dictionary that tracks the TODO items.  If you have a sequence of operations like the following:
1. Update items
2. Trigger a render

You think you're OK because the render happened last right? Well what if a render was happening while you did step 1? This would crash with an unpredicatable error because dictionary is not thread safe. So you need a lock around all property access and rendering. 

It's never safe to access state that's needed during rendering from a BG thread, and it would be gross to use a lock to protect it. The most reasonable thing is to dispatch to the FG thread and let the sync context act-as a zero-overhead lock.

What's worse is that this *might* succeed. It might work pretty well most of the time, but when it fails the error is total nonsense, or it leads to data loss. In the UI frameworks that protect against this explicitly, you at least get a really clear failure message.

To be clear the bug is in step 1. You can't update the state of a component from a background thread, which might mean that an explicit dispatching mechanism *is* needed in these scenarios. It's OK to do step 2 (notify the renderer to do work), but that's the only thing that's safe.

---

I think I need to know a little more about this, I'm hesitating because I'm still not totally sure why this is needed.

Reply:
So to paraphrase, it sounds like you’re saying: it’s not enough to serialise rendering on the sync context, you have to be able to serialise *other* (non-rendering) operations on that same sync context too, to have a way to guarantee they don’t run concurrently with rendering. 

That totally makes sense. With this in mind, it seems inevitable that we need a “Dispatch” API. It’s awkward that developers will only need it for server rendering, but we should be able to report clear errors that occur deterministically if you fail to use it when you must. 

Sounds right?

Reply:
Of course it’s more complex still if you’re mutating state shared by multiple users. Dispatch isn’t enough there (because the sync context is per user) - you really need not to be reading non-thread-safe shared state in rendering at all. For example, have the shared state updater return an immutable copy of the data to each reader. 

But that’s a specialised problem and not something we can solve in the framework.

Reply:
> So to paraphrase, it sounds like you’re saying: it’s not enough to serialise rendering on the sync context, you have to be able to serialise other (non-rendering) operations on that same sync context too, to have a way to guarantee they don’t run concurrently with rendering.

Yes exactly.

Reply:
> Of course it’s more complex still if you’re mutating state shared by multiple users. Dispatch isn’t enough there (because the sync context is per user) - you really need not to be reading non-thread-safe shared state in rendering at all. For example, have the shared state updater return an immutable copy of the data to each reader.

It's also much less problematic to solve. You can make an immutable data structure, or you could use a lock since it's your own code. Trying to manually synchronize your code with the framework is what's hard.

Reply:
> Do we know the cases where this happens? Did user code initiate a background operation? Or do we have a bug that leads to AddToRenderQueue being called from outside the sync context?

It's only user code. Since we eliminated the `ContinueWith` from our lifecycle methods, we no longer have a bug whereby we drop the sync context in the framework AFAIK.